lib/Commands/InterruptSignalAwaiter.h

```diff
@@ -0,0 +1,65 @@
+//===- InterruptSignalHandler.h ---------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2015 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a wrapper for handling signals and processing arrived
+// signals.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLBUILD_COMMAND_INTERRUPTSIGNALAWAITER_H
+#define LLBUILD_COMMAND_INTERRUPTSIGNALAWAITER_H
+
+#include "llbuild/Basic/SignalHandler.h"
+
+#include <atomic>
+#include <functional>
+#include <signal.h>
+#include <thread>
+
+namespace llbuild {
+namespace command {
+struct InterruptSignalAwaiter {
+private:
+  basic::SignalHandler handler{SIGINT, &sigintHandler};
+  static int signalWatchingPipe[2];
+  static std::atomic<bool> wasInterrupted;
+  std::function<void()> interruptHandler;
```

Comment:
I see now that this type is closer to the thing I  had imagined -- I would probably just merge the signal handler method with this code since it should end up being the only thing which interacts with the OS level facility.